core/src/p2p.rs

```diff
@@ -0,0 +1,741 @@
+    pub(crate) read_thread_data: Mutex<ReadThreadData>,
+}
+
+pub(crate) struct ReadThreadData {
+    poll_network_index: u32,
+    pub(crate) packet_cache: Vec<Option<NetworkMessage>>,
+}
+
+pub(crate) enum NetworkMessageVariant {
+    Sumeragi,
+    BlockSync,
+}
+
+#[derive(Debug)]
+pub(crate) struct PeerConnection {
+    stream: TcpStream,
+    crypto: Cryptographer,
+    last_connection_check: Instant,
+}
+
+impl std::fmt::Debug for P2PSystem {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::result::Result<(), std::fmt::Error> {
+        write!(f, "P2PSystem")
+    }
+}
+
+#[allow(clippy::expect_used)]
+impl P2PSystem {
+    #[inline]
+    /// Constructor for [`Self`]
+    pub fn new(listen_addr: String, public_key: PublicKey) -> P2PSystem {
+        P2PSystem {
+            listen_addr,
+            public_key,
+            connect_peer_target: Mutex::new(Vec::new()),
+            connected_to_peers: Mutex::new(HashMap::new()),
+            read_thread_data: Mutex::new(ReadThreadData {
+                poll_network_index: 0_u32,
+                packet_cache: Vec::new(),
+            }),
+        }
+    }
+
+    #[allow(clippy::unwrap_in_result)]
+    fn try_resolve_address(&self) -> Option<std::net::SocketAddr> {
+        let connected_to_peer_keys = {
+            let mut keys = Vec::new();
+            let mut connected_to_peers = self
+                .connected_to_peers
+                .lock();
+            for (public_key, _) in connected_to_peers.iter_mut() {
+                keys.push(public_key.clone());
+            }
+            keys
+        };
+        let target_addrs: Vec<String> = self
+            .connect_peer_target
+            .lock()
+            .iter()
+            .filter(|peer_id| {
+                !connected_to_peer_keys
+                    .iter()
+                    .any(|key| key == &peer_id.public_key)
+            })
+            .map(|peer_id| peer_id.address.clone())
+            .collect();
+        if target_addrs.is_empty() {
+            None
+        } else {
+            let address_candidate = &target_addrs[rand::random::<usize>() % target_addrs.len()];
+            // to_socket_addrs enables dns look-ups.
+            let maybe_addr = address_candidate
+                .to_socket_addrs()
+                .unwrap_or_else(|error| {
+                    debug!(?error, "Socket addr");
+                    Vec::new().into_iter()
+                })
+                .next();
+            if maybe_addr.is_none() {
+                error!(%address_candidate, "Error can't produce address from str. ");
+            }
+            maybe_addr
+        }
+    }
+
+    pub(crate) fn post_to_network(&self, message: &NetworkMessage, recipients: &[PublicKey]) {
+        let mut to_disconnect_keys = Vec::new();
+
+        let mut connected_to_peers = self
+            .connected_to_peers
+            .lock();
+        for public_key in recipients.iter() {
```

Comment:
We can do one better if we use `if let`. The surprising thing about this refactor is that we have a ton of opportunities to exercise `match` to its full potential. I've not done so much pattern matching since Uni.